src/util/simplify_expr.cpp

```diff
@@ -272,91 +272,79 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
   return true;
 }
 
-bool simplify_exprt::simplify_typecast(exprt &expr)
+simplify_exprt::resultt<>
+simplify_exprt::simplify_typecast(const typecast_exprt &expr)
 {
-  if(expr.operands().size()!=1)
-    return true;
-
   const typet &expr_type = expr.type();
-  const typet &op_type = expr.op0().type();
+  const typet &op_type = expr.op().type();
 
   // eliminate casts of infinity
-  if(expr.op0().id()==ID_infinity)
+  if(expr.op().id() == ID_infinity)
   {
     typet new_type=expr.type();
-    exprt tmp;
-    tmp.swap(expr.op0());
+    exprt tmp = expr.op();
     tmp.type()=new_type;
-    expr.swap(tmp);
-    return false;
+    return tmp;
   }
 
   // casts from NULL to any integer
-  if(op_type.id()==ID_pointer &&
-     expr.op0().is_constant() &&
-     to_constant_expr(expr.op0()).get_value()==ID_NULL &&
-     (expr_type.id()==ID_unsignedbv || expr_type.id()==ID_signedbv) &&
-     config.ansi_c.NULL_is_zero)
+  if(
+    op_type.id() == ID_pointer && expr.op().is_constant() &&
+    to_constant_expr(expr.op()).get_value() == ID_NULL &&
+    (expr_type.id() == ID_unsignedbv || expr_type.id() == ID_signedbv) &&
+    config.ansi_c.NULL_is_zero)
   {
-    exprt tmp=from_integer(0, expr_type);
-    expr.swap(tmp);
-    return false;
+    return from_integer(0, expr_type);
   }
 
   // casts from pointer to integer
   // where width of integer >= width of pointer
   // (void*)(intX)expr -> (void*)expr
-  if(expr_type.id()==ID_pointer &&
-     expr.op0().id()==ID_typecast &&
-     expr.op0().operands().size()==1 &&
-     (op_type.id()==ID_signedbv || op_type.id()==ID_unsignedbv) &&
-     to_bitvector_type(op_type).get_width()>=
-     to_bitvector_type(expr_type).get_width())
-  {
-    exprt tmp=expr.op0().op0();
-    expr.op0().swap(tmp);
-    simplify_typecast(expr); // rec. call
-    return false;
+  if(
+    expr_type.id() == ID_pointer && expr.op().id() == ID_typecast &&
+    expr.op().operands().size() == 1 &&
+    (op_type.id() == ID_signedbv || op_type.id() == ID_unsignedbv) &&
+    to_bitvector_type(op_type).get_width() >=
+      to_bitvector_type(expr_type).get_width())
+  {
+    exprt tmp = expr.op().op0();
+    auto new_expr = expr;
+    new_expr.op() = expr.op().op0();
+    return changed(simplify_typecast(new_expr)); // rec. call
   }
 
   // eliminate redundant typecasts
-  if(expr.type() == expr.op0().type())
+  if(expr.type() == expr.op().type())
   {
-    exprt tmp;
-    tmp.swap(expr.op0());
-    expr.swap(tmp);
-    return false;
+    return expr.op();
   }
 
   // eliminate casts to proper bool
   if(expr_type.id()==ID_bool)
   {
     // rewrite (bool)x to x!=0
     binary_relation_exprt inequality(
-      expr.op0(),
+      expr.op(),
       op_type.id() == ID_floatbv ? ID_ieee_float_notequal : ID_notequal,
       from_integer(0, op_type));
     inequality.add_source_location()=expr.source_location();
     simplify_node(inequality);
-    expr.swap(inequality);
-    return false;
+    return inequality;
```

Comment:
Unless I'm mistaken this needs a lot more use of `std::move` at about any return statement?